### PR TITLE
[Core] Adding Parameters constructor to builder and solvers

### PR DESCRIPTION
**Pull request by loumalouomega**
As part of #3185

### CODE REVIEW

**Comment by loumalouomega** [comment]
@roigcarlo Ping

**Review by pooyan-dadvand — DISMISSED** [review]
Why you should pass the linear solver to the constructor with parameter? you may create it with Parameters. Isn't it? 

@RiccardoRossi what is your opinion?

**Comment by RiccardoRossi** [comment]
now that we have the factory for linearsolvers it is best to create it

**Comment by RiccardoRossi** [comment]
now that we have the factory for linearsolvers it is best to create it

**Comment by loumalouomega** [comment]
@RiccardoRossi I don't understand what you mean. The current implementation is an extension of the factories of the linear solvers to the other solving strategies, but this is not working as @pooyan-dadvand expected and duplicates a lot of code. I propose a system to reduce code duplication and work as @pooyan-dadvand expects

**Comment by loumalouomega** [comment]
The problem of creating the linear solver with parameters (which was the original implementation) is that the strategy uses a linear solver too, and it is expected to build it with parameters too. If we construct two different linear solver objects this will cause problems in case we want to interact with the solver (only one of them will be affected). the best way to avoid this problem is construct the linear solver in the strategy and pass it then to the builder and solver

**Comment by RiccardoRossi** [comment]
makes sense

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h [hunk 1]
```diff
@@ -113,11 +113,22 @@ class ResidualBasedBlockBuilderAndSolver
...
 
-    /** Constructor.
+    /**
+     * @brief Default constructor. (with parameters)
+     */
+    explicit ResidualBasedBlockBuilderAndSolver(
+        typename TLinearSolver::Pointer pNewLinearSystemSolver,
+        Parameters ThisParameters
+        ) : BaseType(pNewLinearSystemSolver, ThisParameters)
+    {
```
Comment (RiccardoRossi) [inline kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h:123]: no validation?

File: kratos/solving_strategies/builder_and_solvers/builder_and_solver.h [hunk 2]
```diff
@@ -206,9 +205,21 @@ class BuilderAndSolver
      */
     /*@{ */
 
+    /**
+     * @brief Default constructor. (with parameters)
+     */
+    explicit BuilderAndSolver(typename TLinearSolver::Pointer pNewLinearSystemSolver, Parameters ThisParameters)
```
Comment (RiccardoRossi) [inline kratos/solving_strategies/builder_and_solvers/builder_and_solver.h:211]: validation?

File: kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver_componentwise.h [hunk 3]
```diff
@@ -128,14 +127,25 @@ class ResidualBasedEliminationBuilderAndSolverComponentwise
...
+
+    /**
+     * @brief Default constructor. (with parameters)
+     */
+    explicit ResidualBasedEliminationBuilderAndSolverComponentwise(
+        typename TLinearSolver::Pointer pNewLinearSystemSolver,
+        Parameters ThisParameters
+        ) : ResidualBasedEliminationBuilderAndSolver< TSparseSpace,TDenseSpace,TLinearSolver >(pNewLinearSystemSolver, ThisParameters)
+        , rVar(KratosComponents<TVariableType>::Get(ThisParameters["components_wise_variable"].GetString()))
+    {
```
Comment (RiccardoRossi) [inline kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver_componentwise.h:142]: validation

File: kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver_slip.h [hunk 4]
```diff
@@ -130,12 +129,22 @@ class ResidualBasedEliminationBuilderAndSolverSlip
...
+
+    /**
+     * @brief Default constructor. (with parameters)
      */
-    /*@{ */
+    explicit ResidualBasedEliminationBuilderAndSolverSlip(
+        typename TLinearSolver::Pointer pNewLinearSystemSolver,
+        Parameters ThisParameters
+        ) : ResidualBasedEliminationBuilderAndSolver< TSparseSpace, TDenseSpace, TLinearSolver >(pNewLinearSystemSolver, ThisParameters)
+    {
```
Comment (RiccardoRossi) [inline kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver_slip.h:145]: validation

**Review by RiccardoRossi — CHANGES_REQUESTED** [review]
blocking as all validations need to be implemented.

also removing the FastPR tag. this is definitely NOT a fast pr

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h (hunk 1 above)
Comment (loumalouomega) [reply kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h:123]: I will not force the validation as the parameters that can be passed can contain many things, personally I am against

File: kratos/solving_strategies/builder_and_solvers/builder_and_solver.h (hunk 2 above)
Comment (loumalouomega) [reply kratos/solving_strategies/builder_and_solvers/builder_and_solver.h:211]: Idem

File: kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver_componentwise.h (hunk 3 above)
Comment (loumalouomega) [reply kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver_componentwise.h:142]: Idem

File: kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver_slip.h (hunk 4 above)
Comment (loumalouomega) [reply kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver_slip.h:145]: Idem

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h (hunk 1 above)
Comment (loumalouomega) [reply kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h:123]: For example @RiccardoRossi look at the constructor, calls the base class, which could have a different validator, then throwing an error. That's why there is no validation

**Comment by loumalouomega** [comment]
@RiccardoRossi I set the values where I think makes sense to set the defaults, and adding a note where I think not and why

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h (hunk 1 above)
Comment (RiccardoRossi) [reply kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h:123]: The validator of this class wins over the one of the base class. If the parameters are different then you will have to pass proper parameters to the base class

i consider this as very important

Comment (RiccardoRossi) [reply kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h:123]: If parameters are not compatible, you will have to construct the base class using a constructor that does not pass parameters

Comment (loumalouomega) [reply kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h:123]: OK

File: kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver.h [hunk 5]
```diff
@@ -118,9 +118,9 @@ class ResidualBasedEliminationBuilderAndSolver
     explicit ResidualBasedEliminationBuilderAndSolver(
         typename TLinearSolver::Pointer pNewLinearSystemSolver,
         Parameters ThisParameters
-        ) : BaseType(pNewLinearSystemSolver, ThisParameters)
+        ) : BaseType(pNewLinearSystemSolver)
     {
-        // NOTE: We don't validate because Parameters are not used, additionally this is the base class of others B&S and may case false errors if any default is check
+        // NOTE: We don't validate because Parameters are not used
```
Comment (RiccardoRossi) [inline kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver.h]: I am sorry but this does not make sense...

File: kratos/solving_strategies/builder_and_solvers/builder_and_solver.h [hunk 6]
```diff
@@ -206,9 +205,24 @@ class BuilderAndSolver
...
+    {
+        // NOTE: We don't validate because Parameters are not used, additionally this is the base class of others B&S and may case false errors if any default is check
```
Comment (RiccardoRossi) [inline kratos/solving_strategies/builder_and_solvers/builder_and_solver.h]: the point is actually that if someone tries to pass a parameter that cannot be accepted an error should be thrown.

In this case you should check that no parameters are passed

if you have a derived class, then you will do the validation in the derived class and construct the base class by using the parameters without validation.

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_constraints.h [hunk 7]
```diff
@@ -103,15 +103,28 @@ class ResidualBasedBlockBuilderAndSolverWithConstraints
...
-    /** Constructor.
+    /**
+     * @brief Default constructor. (with parameters)
+     */
+    explicit ResidualBasedBlockBuilderAndSolverWithConstraints(
+        typename TLinearSolver::Pointer pNewLinearSystemSolver,
+        Parameters ThisParameters
+        ) : BaseType(pNewLinearSystemSolver, ThisParameters)
+    {
+        // NOTE: We don't validate because Parameters are not used
```
Comment (RiccardoRossi) [inline kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_constraints.h]: you need to validate to ensure that no parameters can be passed and ignored

File: kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver_slip.h (hunk 4 above)
Comment (RiccardoRossi) [reply kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver_slip.h:145]: here you should validate the parameters (which btw, cannot be empty here) and pass constructor the base class using the constructor without parameters

**Comment by loumalouomega** [comment]
@RiccardoRossi I addressed all your comments

File: kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver_componentwise.h [hunk 8]
```diff
@@ -128,16 +128,35 @@ class ResidualBasedEliminationBuilderAndSolverComponentwise
...
+     */
+    explicit ResidualBasedEliminationBuilderAndSolverComponentwise(
+        typename TLinearSolver::Pointer pNewLinearSystemSolver,
+        Parameters ThisParameters
+        ) : ResidualBasedEliminationBuilderAndSolverType(pNewLinearSystemSolver)
+    {
+        // Validate default parameters
+        Parameters default_parameters = Parameters(R"(
+        {
+            "components_wise_variable" : "DISPLACEMENT"
```
Comment (RiccardoRossi) [inline kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver_componentwise.h]: DISPLACEMENT is incorrect as a componentwise variable...

it should be either a scalar variable (PRESSURE) or a component DISPLACEMENT_X

in any case i would not put a default but rather ask by default for something like "SCALAR_VARIABLE_OR_COMPONENT"

Comment (loumalouomega) [reply kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver_componentwise.h]: OK

**Comment by loumalouomega** [comment]
Done

File: kratos/solving_strategies/builder_and_solvers/builder_and_solver.h [hunk 9]
```diff
@@ -206,9 +205,29 @@ class BuilderAndSolver
...
+        mReshapeMatrixFlag = false;
+    }
+
     /** Constructor.
      */
-    BuilderAndSolver(
+    explicit BuilderAndSolver(
         typename TLinearSolver::Pointer pNewLinearSystemSolver)
     {
         mpLinearSystemSolver = pNewLinearSystemSolver;
```
Comment (RiccardoRossi) [inline kratos/solving_strategies/builder_and_solvers/builder_and_solver.h:233]: not essential, but you could call here the constructor by parameters

**Review by RiccardoRossi — APPROVED** [review]

File: kratos/solving_strategies/builder_and_solvers/builder_and_solver.h (hunk 9 above)
Comment (loumalouomega) [reply kratos/solving_strategies/builder_and_solvers/builder_and_solver.h:233]: For the linear solver?, no, as I said the Linear solver is already constructed in the Strategy, if you do that you will create two linear solvers, which is not very consistent

Comment (RiccardoRossi) [reply kratos/solving_strategies/builder_and_solvers/builder_and_solver.h:233]: you could have done
 
     explicit BuilderAndSolver(
        typename TLinearSolver::Pointer pNewLinearSystemSolver)
     {
            BuilderAndSolver(pNewLinearSystemSolver, Parameters() )
     }

and essentially implement one single constructor

OR EVEN just define one single constructor passing by default an empty parametes object

Comment (loumalouomega) [reply kratos/solving_strategies/builder_and_solvers/builder_and_solver.h:233]: Huuum, at the end is the same thing. Can be changed in the future

**Pull request merged**